lib/migrations/base.ts

```diff
@@ -0,0 +1,210 @@
+          storageKey,
+          PreviousSnjsVersion1_0_0
+        );
+      } else {
+        /** Coming from 2.0.0 (which did not store version) OR is brand new application */
+        const migrationKey = namespacedKey(this.services!.identifier, LastMigrationTimeStampKey2_0_0);
+        const migrationValue = await this.services.deviceInterface.getRawStorageValue(migrationKey);
+        const is_2_0_0_application = !isNullOrUndefined(migrationValue);
+        if (is_2_0_0_application) {
+          await this.services.deviceInterface.setRawStorageValue(
+            storageKey,
+            PreviousSnjsVersion2_0_0
+          );
+          await this.services.deviceInterface.removeRawStorageValue(LastMigrationTimeStampKey2_0_0);
+        } else {
+          /** Is new application, use current version as not to run any migrations */
+          await this.services.deviceInterface.setRawStorageValue(
+            storageKey,
+            SnjsVersion
+          );
+        }
+      }
+    }
+  }
+
+  /**
+   * If the keychain is empty, and the user does not have a passcode,
+   * AND there appear to be stored account key params, this indicates
+   * a launch where the keychain was wiped due to restoring device
+   * from cloud backup which did not include keychain. This typically occurs
+   * on mobile when restoring from iCloud, but we'll also follow this same behavior
+   * on desktop/web as well, since we recently introduced keychain to desktop.
+   *
+   * We must prompt user for account password, and validate based on ability to decrypt
+   * an item. We cannot validate based on storage because 1.x mobile applications did
+   * not use encrypted storage, although we did on 2.x. But instead of having two methods
+   * of validations best to use one that works on both.
+   *
+   * The item is randomly chosen, but for 2.x applications, it must be an items key item
+   * (since only item keys are encrypted directly with account password)
+   */
+  private async preemptivelyRepairMissingKeychain() {
+    const version = (await this.getStoredVersion())!;
+    const reader = CreateReader(
+      version,
+      this.services.deviceInterface,
+      this.services.identifier,
+      this.services.environment
+    );
+    const usesKeychain = reader.usesKeychain;
+    if (!usesKeychain) {
+      /** Doesn't apply if this version did not use a keychain to begin with */
+      return;
+    }
+
+    const rawAccountParams = await reader.getAccountKeyParams();
+    const hasAccountKeyParams = !isNullOrUndefined(rawAccountParams);
+    if (!hasAccountKeyParams) {
+      /** Doesn't apply if account is not involved */
+      return;
+    }
+
+    const hasPasscode = await reader.hasPasscode();
+    if (hasPasscode) {
+      /** Doesn't apply if using passcode, as keychain would be bypassed in that case */
+      return;
+    }
+
+    const accountKeysMissing = !(await reader.hasNonWrappedAccountKeys());
+    if (!accountKeysMissing) {
+      return;
+    }
```

Comment:
I don't know, the idea is to make the variables describe the intention, in place of a comment. `accountKeysMissing` tells you exactly the problem, but `hasAccountKeys` doesn't tell you what the problem or premise is.

But it could easily go both ways I suppose.